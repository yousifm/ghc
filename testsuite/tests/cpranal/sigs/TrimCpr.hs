{-# LANGUAGE GADTs #-}

module TrimCpr where

expensive :: Int -> Bool
expensive = even
{-# NOINLINE expensive #-}

-- | Should *not* have the CPR property, because it has arity 1.
trimByArity :: Int -> Int -> Maybe Int
trimByArity x
  | expensive x = \y -> Just (x+y)
  | otherwise   = \y -> Just y
{-# NOINLINE trimByArity #-}

data T a where
  A :: T Int
  B :: T (Int -> Int)

-- | If we say `LLb` for 'g', this will get the CPR property, which is wrong.
trimByArity2 :: Int -> Int -> Int
trimByArity2 0 y = y + 1
trimByArity2 x y = g B (x+y)
  where
    g :: T a -> a
    g A = error "A"
    g B = \_ -> error "B"
    {-# NOINLINE g #-}

-- | Should *not* have the CPR property, because evaluation of `blah` won't
-- terminate and `f` won't have the nested CPR property.
trimByTerm :: Int -> Int
trimByTerm n
  | n < 0     = n
  | otherwise = case f n of Just blah -> blah
  where
    f :: Int -> Maybe Int
    f n = Just (sum [0..n])
    {-# NOINLINE f #-}
{-# NOINLINE trimByTerm #-}
