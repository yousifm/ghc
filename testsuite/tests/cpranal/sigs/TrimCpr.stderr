
==================== Term signatures ====================
TrimCpr.$tc'A: *
TrimCpr.$tc'B: *
TrimCpr.$tcT: *
TrimCpr.$trModule: *
TrimCpr.expensive: *
TrimCpr.trimByArity: *L#
TrimCpr.trimByArity2: *1(#)
TrimCpr.trimByTerm: *



==================== Cpr signatures ====================
TrimCpr.$tc'A:
TrimCpr.$tc'B:
TrimCpr.$tcT:
TrimCpr.$trModule:
TrimCpr.expensive:
TrimCpr.trimByArity:
TrimCpr.trimByArity2:
TrimCpr.trimByTerm:


